Tests/NIOHTTP1Tests/HTTPClientUpgradeTests.swift

```diff
@@ -0,0 +1,1022 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2017-2019 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import XCTest
+import Dispatch
+@testable import NIO
+@testable import NIOHTTP1
+
+extension ChannelPipeline {
+    
+    // Waits up to 1 second for the upgrader to be removed by polling the pipeline
+    // every 50ms checking for the handler.
+    fileprivate func waitForClientUpgraderToBeRemoved() throws {
```

Comment:
Rather than use this, it'd be better to write these tests in terms of `EmbeddedChannel`. `EmbeddedChannel` is strictly deterministic, so the tests avoid flakiness and timing sensitivity. They also run faster!

Reply:
This is far more succinct and also removes any dependency on server channels. So an all-round win. It is the first time I have used EmbeddedChannel though. Would you mind giving it a look, once pushed, just to check that you are happy with the usage. Thanks.